### PR TITLE
Sentiment analysis

### PR DESCRIPTION
**Pull request by Mast3rwaf1z**

### CODE REVIEW

File: include/PredictionManager.hpp [hunk 1]
```diff
@@ -86,10 +87,9 @@ class PredictionBuilder {
             final_data.push_back(pair.second ? 1.0 : 0.0);
         }
         // now we normalize
```
Comment (Taoshix) [inline include/PredictionManager.hpp:89]: fake news, it is not normalized, this is false advertisement and you will therefore get sued if this isn't removed

File: include/Routes/Answers.hpp [hunk 2]
```diff
@@ -19,7 +19,7 @@ class Answers : public Route {
             auto answerId = stoi(request.path_params["answerId"]);
             auto token = request.path_params["token"];
             json response_data;
-            if (answerId <= 0 || db->answers->get_where("id", db_int(answerId)).size() == 0) {
+            if (answerId <= 0 || db->answers->get_where("id", answerId).size() == 0) {
```
Comment (Taoshix) [inline include/Routes/Answers.hpp:22]: me like

File: include/Routes/Mitigations.hpp [hunk 3]
```diff
@@ -35,7 +35,7 @@ class Mitigations : public Route {
         /// @brief Returns a specific mitigation with a given id.
         this->server->Get("/mitigations/tags/:tag", [&](Request request, Response& response){
             auto tagparam = request.path_params["tag"];
-            auto tags = split(tagparam);
+            auto tags = P8::split_string(tagparam);
```
Comment (Taoshix) [inline include/Routes/Mitigations.hpp:38]: yeah, i should have done that renaming myself, glad to see that you did it

File: include/Utils.hpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+ public:
+    explicit ThreadPool(int size) {
+        if (size <= 0)
+            throw exception();
+        this->size = size;
+    }
+    vector<Output> map(function<Output(Input)> f, vector<Input> values, function<void(Input, Output)> logger) {
+        vector<Worker<Input, Output>> workers;
```
Comment (Taoshix) [inline include/Utils.hpp:104]: vector vector = new vector type stuff

File: include/Utils.hpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            workers[i%size].add_data(i, value);
+        return map(workers);
+    }
+    vector<Output> map(function<Output(Input)> f, vector<Input> values) {
+        vector<Worker<Input, Output>> workers;
+        for (auto i = 0; i < size; i++)
+            workers.push_back(Worker<Input, Output>(f));
+        for (auto [i, value] : ranges::zip_view(make_range(values.size()), values))
+            workers[i%size].add_data(i, value);
+        return map(workers);
```
Comment (Taoshix) [inline include/Utils.hpp:121]: haskell

File: include/Utils.hpp [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    vector<Output> map(function<Output(Input)> f, vector<Input> values, function<void(Input, Output)> logger) {
+        vector<Worker<Input, Output>> workers;
+
+        for (auto i = 0; i < size; i++) {
+            workers.push_back(Worker<Input, Output>(f));
+        }
+        for (auto& worker : workers)
+            worker.set_logger(logger);
+        for (auto [i, value] : ranges::zip_view(make_range(values.size()), values))
+            workers[i%size].add_data(i, value);
```
Comment (Taoshix) [inline include/Utils.hpp:112]: mod in subscripting wtf

File: include/Utils.hpp [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            workers.push_back(Worker<Input, Output>(f));
+        for (auto [i, value] : ranges::zip_view(make_range(values.size()), values))
+            workers[i%size].add_data(i, value);
+        return map(workers);
+    }
+};
+map<string, string> run_cmd(string command) {
+    map<string, string> result;
+    auto stdout_path = "/tmp/p8_backend_stdout";
+    auto stderr_path = "/tmp/p8_backend_stderr";
```
Comment (Taoshix) [inline include/Utils.hpp:127]: linux exclusive

File: src/backend.cpp [hunk 8]
```diff
@@ -28,6 +35,20 @@ void setup(std::string path) {
...
+    POPULATE
+};
+Mode to_mode(string input) {
+    if (input == "default")
+        return DEFAULT;
+    else if (input == "check")
+        return CHECK;
+    else
+        return POPULATE;
+}
```
Comment (Taoshix) [inline src/backend.cpp:50]: should be case insensitive

File: src/backend.cpp [hunk 9]
```diff
@@ -44,21 +65,97 @@ int main(int argc, char* argv[]) {
         .scan<'d', int>()
         .default_value(8080)
         .nargs(1);
+    program.add_argument("--dataset", "-d")
+        .help("Path to the dataset for populating the database")
+        .default_value("./")
+        .nargs(1);
+    program.add_argument("--threads", "-t")
+        .help("Specify number of threads that can be used")
+        .scan<'d', int>()
```
Comment (Taoshix) [inline src/backend.cpp:74]: should be unsigned int

File: src/backend.cpp [hunk 10]
```diff
@@ -44,21 +65,97 @@ int main(int argc, char* argv[]) {
...
+        .help("Operating mode for the backend")
+        .default_value("default");
 
     try {
         program.parse_args(argc, argv);
     }
-    catch (const std::exception& e) {
-        std::cerr << "argparse pooped itself, exiting..." << std::endl;
+    catch (const exception& e) {
+        cerr << "argparse pooped itself, exiting..." << endl;
```
Comment (Taoshix) [inline src/backend.cpp:85]: true

File: src/backend.cpp [hunk 11]
```diff
@@ -44,21 +65,97 @@ int main(int argc, char* argv[]) {
...
+            break;
+        }
+        case CHECK: {
+            cout << "Checking database contents..." << endl;
+            auto path = program.get<string>("--database");
+            auto thread_cnt = program.get<int>("--threads");
+            Database db{path};
+            auto uids = db.users->get_where();
+            auto uids_size = uids.size();
+            vector<future<string>> tasks;
```
Comment (Taoshix) [inline src/backend.cpp:107]: future<vector<future<string>>> or just remove

File: src/backend.cpp [hunk 12]
```diff
@@ -44,21 +65,97 @@ int main(int argc, char* argv[]) {
...
+                args.push_back(make_pair(i, uid));
+            }
+            auto target = [=](pair<int, int> arg) {
+                auto i = arg.first;
+                auto uid = arg.second;
+                auto jids = db.journals->get_where("userId", uid);
+                vector<double> results;
+                for (auto jid : jids) {
+                    auto aids = db.answers->get_where("journalId", jid);
+                    PredictionManager manager;
```
Comment (Taoshix) [inline src/backend.cpp:120]: should be moved out of loop, this is slowing it down

File: src/backend.cpp [hunk 13]
```diff
@@ -44,21 +65,97 @@ int main(int argc, char* argv[]) {
...
+            break;
+        }
+        case POPULATE: {
+            auto prefix = program.get<string>("--dataset");
+            auto database = program.get<string>("--database");
+            cout << "REMOVING DATABASE IN 10 SECONDS" << endl;
+            for (auto i = 10; i >= 0; i--) {
+                sleep(1);
+                cout << i << endl;
+            }
```
Comment (Taoshix) [inline src/backend.cpp:150]: funny countdown

File: src/backend.cpp [hunk 14]
```diff
@@ -44,21 +65,97 @@ int main(int argc, char* argv[]) {
...
+            auto rm_result = system(format("rm {}", database).c_str());
+            cout << "Creating database" << endl;
+            {
+                auto db = Database{database};
```
Comment (Taoshix) [inline src/backend.cpp:154]: nice constructor destructor

File: src/questionaire_test.cpp [hunk 15]
```diff
@@ -1,2 +0,0 @@
-int main() {
-}
```
Comment (Taoshix) [inline src/questionaire_test.cpp:2]: why you remove my bro

File: src/utils_test.cpp [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+using namespace std;
+
+typedef Test<function<void()>> UtilsTest;
+
+int main() {
+    auto utils = UtilsTest();
+    utils.add_test("Test Make Range: edge cases", [&](){
+        // testing edge cases
+        auto range1 = P8::make_range(0);
+        auto range2 = P8::make_range(-1);
```
Comment (Taoshix) [inline src/utils_test.cpp:15]: ITV moment

File: src/utils_test.cpp [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+        assert(range1.size() == 0);
+        assert(range2.size() == 0);
+    });
+    utils.add_test("Test Make Range: funky types", [&](){
+        auto range = P8::make_range<vector<map<int, string>>>(1, [](int i){
```
Comment (Taoshix) [inline src/utils_test.cpp:21]: could have been even more funky with
future<<vector<map<int, string>>>>

File: src/utils_test.cpp [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+            vector<map<int, string>> result = {
+                {
+                    make_pair(i, "test")
+                }
+            };
+            return result;
+        });
+        vector<vector<map<int, string>>> control = {{{make_pair(0, "test")}}};
```
Comment (Taoshix) [inline src/utils_test.cpp:29]: when auto fails

File: include/Utils.hpp [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+    }
+    return result;
+}
+vector<string> split_string(string s, string delimiter) {
+    vector<string> res;
+    int pos = 0;
+    while (pos < s.size()) {
+        pos = s.find(delimiter);
+        res.push_back(s.substr(0, pos));
+        s.erase(0, pos+1);
```
Comment (Taoshix) [inline include/Utils.hpp:147]: +1 should be be length of delimiter
it was +1 because of comma is only a single character

**Review by Taoshix — CHANGES_REQUESTED** [review]
few changes and lots of comments

**Review by Taoshix — APPROVED** [review]
Appears to be good now compared to first iteration
